rust/serving/src/source.rs

```diff
@@ -0,0 +1,269 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use bytes::Bytes;
+use tokio::sync::{mpsc, oneshot};
+use tokio::time::Instant;
+
+use crate::app::callback::state::State as CallbackState;
+use crate::app::callback::store::redisstore::RedisConnection;
+use crate::app::tracker::MessageGraph;
+use crate::Settings;
+use crate::{Error, Result};
+
+/// [Message] with a oneshot for notifying when the message has been completed processed.
+pub(crate) struct MessageWrapper {
+    // TODO: this might be more that saving to ISB.
+    pub(crate) confirm_save: oneshot::Sender<()>,
+    pub(crate) message: Message,
+}
+
+/// Serving payload passed on to Numaflow.
+#[derive(Debug)]
+pub struct Message {
+    pub value: Bytes,
+    pub id: String,
+    pub headers: HashMap<String, String>,
+}
+
+enum ActorMessage {
+    Read {
+        batch_size: usize,
+        timeout_at: Instant,
+        reply_to: oneshot::Sender<Vec<Message>>,
+    },
+    Ack {
+        offsets: Vec<String>,
+        reply_to: oneshot::Sender<()>,
+    },
+}
+
+struct ServingSourceActor {
+    messages: mpsc::Receiver<MessageWrapper>,
+    handler_rx: mpsc::Receiver<ActorMessage>,
+    tracker: HashMap<String, oneshot::Sender<()>>,
+}
+
+impl ServingSourceActor {
+    async fn start(
+        settings: Arc<Settings>,
+        handler_rx: mpsc::Receiver<ActorMessage>,
+    ) -> Result<()> {
+        let (messages_tx, messages_rx) = mpsc::channel(10000);
+        // Create a redis store to store the callbacks and the custom responses
+        let redis_store = RedisConnection::new(settings.redis.clone()).await?;
+        // Create the message graph from the pipeline spec and the redis store
+        let msg_graph = MessageGraph::from_pipeline(&settings.pipeline_spec).map_err(|e| {
+            Error::InitError(format!(
+                "Creating message graph from pipeline spec: {:?}",
+                e
+            ))
+        })?;
+        let callback_state = CallbackState::new(msg_graph, redis_store).await?;
+
+        tokio::spawn(async move {
+            let mut serving_actor = ServingSourceActor {
+                messages: messages_rx,
+                handler_rx,
+                tracker: HashMap::new(),
+            };
+            serving_actor.run().await;
+        });
+        let app = crate::AppState {
+            message: messages_tx,
+            settings,
+            callback_state,
+        };
+        tokio::spawn(async move {
+            crate::serve(app).await.unwrap();
+        });
```

Comment:
make sure these tokio tasks are aborted during shutdown, structured concurrency :).

Reply:
This might need more work. Created an issue for it https://github.com/numaproj/numaflow/issues/2309